components/select/Select.razor.cs

```diff
@@ -84,33 +86,29 @@ public IEnumerable<TItem> DataSource
                     SelectedOptionItems.Clear();
 
                     Value = default;
-                    var sameObject = object.ReferenceEquals(_datasource, value);
 
                     _datasource = value;
 
-                    if (!sameObject)
-                        OnDataSourceChanged?.Invoke();
+                    OnDataSourceChanged?.Invoke();
 
                     return;
                 }
 
                 if (value != null)
                 {
-                    bool hasChanged;
-
                     if (_datasource == null)
                     {
-                        hasChanged = true;
+                        _dataSourceHasChanged = true;
                     }
                     else
                     {
-                        hasChanged = !value.SequenceEqual(_datasource);
+                        _dataSourceHasChanged = !value.SequenceEqual(_datasource)
+                            || (_optionsHasInitialized && SelectOptionItems.Count != _datasource.Count());
```

Comment:
I don't think we can rely on the item count, and `SelectOptionItems` also contains added tags?

Reply:
I forgot about the tags. Indeed it does, but it can be simply fixed by subtracting what we have in `AddedTags`.
Count gives a basic check over the basic add/remove. It obviously won't work when a consumer adds & remove the same number of items from the `DataSource`. From what I can see, currently the whole sequence comparison in `DataSource` setter is useless anyway. We could just replace it with reference equality check and I think it would work the same (except probably faster).
But, we should probably keep the `SequenceEqual` check but not to check it against `_dataSource` but against a deep copy of the `_dataSource`. I think then we could safely rely on the change detection.

Reply:
Sorry, I think shallow copy should be good enough.

Reply:
My tests revealed that shallow copy wont cover all scenarios (given the `DataSource` is an `IEnumerable<Persons>`):
```csharp
        //will NOT detect
        _persons[2].Name = newValue; 
...
        //will detect
        _persons[2] = new Person { Id = 3, Name = "new Person()"  }; 
...
        //will detect
        _persons.Remove(_persons.Last());
        _persons.Add(new Person { Id = 3, Name = "removed and added" });
```
But we cannot really make a deep copy ourselves (possible issues: cycle object graph, references to external objects, events). So, to cover all scenarios, the only other way I can think of is to allow the consumers to pass their own `IEqualityComparer` - I would think these are going to be the advanced scenarios, but still available to the developers.

Reply:
It seems we are breaking this use-case: https://ant-design-blazor.gitee.io/zh-CN/components/select#components-select-demo-TrackItemChanges. Now a custom IEqualityComparer is necessary to track item changes.

Reply:
Thanks, I will have a look. 

I also experimented with `IEqualityComparer` but I think it won't really help us here, because in case of changes to an item of the `DataSource` we still are making changes to the same object. So the comparer compares an object to itself. Now I start to think that we should expose our own interface/base class for `DataSource` that will have a `Clone` virtual method, which will do a shallow copy by default but with ability to overload it, so the consumers can create a deep copy implementation. What do you think?